src/Model/LazyBSONArray.php

```diff
@@ -0,0 +1,325 @@
+
+    /** @var array<int, TValue> */
+    private array $set = [];
+
+    /** @var array<int, true> */
+    private array $unset = [];
+
+    private bool $entirePackedArrayRead = false;
+
+    private CodecLibrary $codecLibrary;
+
+    /**
+     * Deep clone this lazy array.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            $this->set[$key] = recursive_copy($value);
+        }
+    }
+
+    /** @param PackedArray<TValue>|list<TValue>|null $input */
+    public function __construct($input = null, ?CodecLibrary $codecLibrary = null)
+    {
+        if ($input === null) {
+            $this->bson = PackedArray::fromPHP([]);
+        } elseif ($input instanceof PackedArray) {
+            $this->bson = $input;
+        } elseif (is_array($input)) {
+            if (! array_is_list($input)) {
+                // Check if all keys are numeric
+                foreach ($input as $key => $value) {
+                    if (! is_numeric($key)) {
+                        throw InvalidArgumentException::invalidType('input', $input, [PackedArray::class, 'array', 'null']);
+                    }
+                }
+            }
+
+            $this->bson = PackedArray::fromPHP([]);
+            $this->set = $input;
+            $this->exists = array_map(
+                /** @param TValue $value */
+                fn ($value): bool => true,
+                $this->set,
+            );
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [PackedArray::class, 'array', 'null']);
+        }
+
+        $this->codecLibrary = $codecLibrary ?? new LazyBSONCodecLibrary();
+    }
+
+    /** @return array{bson: PackedArray<TValue>, set: array<int, TValue>, unset: array<int, true>, codecLibrary: CodecLibrary} */
+    public function __serialize(): array
+    {
+        return [
+            'bson' => $this->bson,
+            'set' => $this->set,
+            'unset' => $this->unset,
+            'codecLibrary' => $this->codecLibrary,
+        ];
+    }
+
+    /** @param array{bson: PackedArray<TValue>, set: array<int, TValue>, unset: array<int, true>, codecLibrary: CodecLibrary} $data */
+    public function __unserialize(array $data): void
+    {
+        $this->bson = $data['bson'];
+        $this->set = $data['set'];
+        $this->unset = $data['unset'];
+        $this->codecLibrary = $data['codecLibrary'];
+
+        $this->exists = array_map(
+        /** @param TValue $value */
+            fn ($value): bool => true,
+            $this->set,
+        );
+
+        foreach ($this->unset as $index => $unused) {
+            $this->exists[$index] = false;
+        }
+    }
+
+    public function count(): int
+    {
+        $this->readEntirePackedArray();
+
+        return count(array_filter($this->exists));
+    }
+
+    /** @return ListIterator<TValue> */
+    public function getIterator(): ListIterator
```

Comment:
Seeing more weird spacing in the GitHub PR view:

```suggestion
    public function getIterator(): ListIterator
```